docs/reference/fields.txt

```diff
@@ -576,93 +578,341 @@ you have three options:
+    def mongoize
+      [ x, y ]
+    end
 
-.. code-block:: ruby
+    class << self
 
-  class Order
-    include Mongoid::Document
+      # Takes any possible object and converts it to how it would be
+      # stored in the database.
+      def mongoize(object)
+        case object
+        when Point then object.mongoize
+        when Hash then Point.new(object[:x], object[:y]).mongoize
+        else object
+        end
+      end
 
-    field :state, type: String, default: 'created'
+      # Get the object as it was stored in the database, and instantiate
+      # this custom class from it.
+      def demongoize(object)
+        Point.new(object[0], object[1])
+      end
+
+      # Converts the object that was supplied to a criteria and converts it
+      # into a query-friendly form.
+      def evolve(object)
+        case object
+        when Point then object.mongoize
+        else object
+        end
+      end
+    end
   end
 
-The default value can also be specified as a ``Proc``:
+The instance method ``mongoize`` takes an instance of your custom type object, and
+converts it into a represenation of how it will be stored in the database, i.e. to pass
+to the MongoDB Ruby driver. In our example above, we want to store our ``Point``
+object as an ``Array`` in the form ``[ x, y ]``.
+
+The class method ``mongoize`` is similar to the instance method, however it must handle
+objects of all possible types as inputs. The ``mongoize`` method is used when calling the
+setter methods for fields of your custom type.
 
 .. code-block:: ruby
 
-  class Order
-    include Mongoid::Document
+   point = Point.new(12, 24)
+   venue = Venue.new(location: point) # This uses the Point#mongoize instance method.
+   venue = Venue.new(location: [ 12, 24 ]) # This uses the Point.mongoize class method.
 
-    field :fulfill_by, type: Time, default: ->{ Time.now + 3.days }
-  end
+The class method ``demongoize`` does the inverse of ``mongoize``. It takes the raw object
+from the MongoDB Ruby driver and converts it to an instance of your custom type.
+In this case, the database driver returns an ``Array`` and we instantiate a ``Point`` from it.
+The ``demongoize`` method is used when calling the getters of fields for your custom type.
+Note that in the example above, since ``demongoize`` calls ``Point.new``, a new instance of
+``Point`` will be generated on each call to the getter.
 
 .. note::
 
-  Default values that are not ``Proc`` instances are evaluated at class load
-  time, meaning the following two definitions are not equivalent:
+  The ``mongoize`` and ``demongoize`` methods should return ``nil`` on values
+  that are uncastable to your custom type. See the section on :ref:`Uncastable
+  Values <uncastable-values>` for more details.
 
-  .. code-block:: ruby
+Lastly, the class method ``evolve`` is similar to ``mongoize``, however it is used
+when transforming objects for use in Mongoid query criteria.
```

Comment:
You should note that the difference between evolve and mongoize is that it doesn't return nil on uncastable values.

Reply:
Added this to https://jira.mongodb.org/browse/MONGOID-5460.